### ISSUE TITLE
Increase the number of dims for KNN vectors to 2048 [LUCENE-10471]

### ISSUE THREAD

**Issue by asfimport** [apache/lucene#11507]
The current maximum allowed number of dimensions is equal to 1024. But we see in practice a couple well-known models that produce vectors with > 1024 dimensions (e.g [mobilenet_v2](https://tfhub.dev/google/imagenet/mobilenet_v2_035_224/feature_vector/1) uses 1280d vectors, OpenAI / GPT-3 Babbage uses 2048d vectors). Increasing max dims to `2048` will satisfy these use cases.

I am wondering if anybody has strong objections against this.

---
Migrated from [LUCENE-10471](https://issues.apache.org/jira/browse/LUCENE-10471) by Mayya Sharipova (@mayya-sharipova), 6 votes, updated Aug 15 2022
Pull requests: https://github.com/apache/lucene/pull/874

**Comment by aykutfirat** [comment]
Lots of things happened since Aug, like the arrival of ChatGPT, and people's increased desire to use OpenAI's state of the art embeddings which are of size 1536. Can you at least please increase it to 2048 for now, while you discuss upper limits?

**Comment by uschindler** [comment]
> Lots of things happened since Aug, like the arrival of ChatGPT, and people's increased desire to use OpenAI's state of the art embeddings which are of size 1536. Can you at least please increase it to 1536 for now, while you discuss upper limits?

Actually it is a one-line change (without any garantees), see https://github.com/apache/lucene/pull/874/files

If you really want to shoot you in the foot: Download source code of Lucene in the version you need for your Elasticsearch instance (I assume you coming from https://github.com/elastic/elasticsearch/issues/92458), patch it  with #874, and then run './gradlew distribution'. Copy the JAR files into your ES districution. Done.

But it is not sure if this will blos up and indexes created by that won't read anymore with standard Lucene

**Comment by uschindler** [comment]
Why I was doing that suggestion: If you are interested, try it out with your dataset and your Elasticsearch server and report back! Maybe you will figure out that performance does not work or memory usage is too high.

**Comment by gibrown** [comment]
I'll preface this by saying I am also skeptical that going beyond 1024 makes sense for most use cases and scaling is a concern. However, amidst the current excitement to try and use openai embeddings the first cut at choosing a system to store and use those embeddings was Elasticsearch. Then the 1024 limit was run into and so various folks are looking at other alternatives largely because of this limit.

The use cases tend to be Q/A, summarization, and recommendation systems for WordPress and Tumblr. There are multiple proof of concept systems people have built (typically on top of various typscript, javascript, or python libs) which use the openai embeddings directly (and give quite impressive results). Even though I am pretty certain that reducing the dimensions will be a better idea for many of these, the ability to build and prototype on higher dimensions would be extremely useful.

**Comment by FranciscoBorges** [comment]
@uschindler @rmuir FWIW We are interested in using Lucene's kNN with 1536 dimensions in order to use OpenAI's embeddings API. We benchmarked a patched Lucene/Solr. We fully understand (we measured it :-P) that there is an increase in memory consumption and latency. Sure thing.

We have applications where dev teams have chosen to work with OpenAI embeddings and where the number of records involved and requests per second make the trade offs of memory and latency perfectly acceptable.

There is a great deal of enthusiasm around OpenAI and releasing a working application ASAP. For many of these the resource cost of 1536 dimensions is perfectly acceptable against the alternative of delaying a pilot to optimize further.

Our work would be a lot easier if Lucene's kNN implementation supported 1536 dimensions without need for a patch.

**Comment by dsmiley** [comment]
I'm reminded of the great maxBooleanClauses debate.  At least that limit is user configurable (for the system deployer; not the end user doing a query) whereas this new one for kNN is not.

I can understand how we got to this point -- limits often start as hard limits.  The current limit even seems high based on what has been said.  But users have spoken here on a need for them to _configure_ Lucene for _their_ use-case (such as experimentation within a system they are familiar with) _and accept the performance consequences_.  **I would like this to be possible with a System property**.  This hasn't been expressly asked?  Why should Lucene, just a library that doesn't know what's best for the user, prevent a user from being able to do that?  

This isn't an inquiry about why limits exist; of course systems need limits.

**Comment by alessandrobenedetti** [comment]
Hi @dsmiley  I updated the dev discussion on the mailing list:
[Proposal] Remove max number of dimensions for KNN vectors

And proceeded with a pragmatic new mail thread, where we just collect proposals with a motivation (no discussion there):
Dimensions Limit for KNN vectors - Next Steps

Feel free to participate!
My intention is to act relatively fast (and then also operate Solr side).
It's a train we don't need/want to miss!

**Comment by ryantbrown** [comment]
The rabbit hole that is trying to store Open AI embeddings in Elasticsearch eventually leads here.  I read the entire thread and unless I am missing something, the obvious move to make the limit configurable (up to a point) or at a minimum, increase the limit to 1536 to support the `text-embedding-ada-002` model. In other words, there should be a compelling reason _not_ to increase the limit beyond the fact that it will hard to reduce in the future.

**Comment by nknize** [comment]
Cross posting here because I responded to the PR instead of this issue.

> ...why is it then that GPT-4, which internally represents each token with a vector of more than 8192, still inaccurately recalls information about entities?

I think this comment actually supports @MarcusSorealheis argument?  e.g., What's the point in indexing 8K dimensions if it isn't much better at recall than 768?


> If the real issue is with the use of HNSW, which isn't suitable for this, not that highe-dimensionality embeddings have value, then the solution isn't to not provide the feature, but to switch technologies to something more suitable for the type of applications that people use Lucene for.

I may be wrong but it seems like this is where most of the lucene committers here are settling? 

Over a decade ago I wanted a high dimension index for some facial recognition and surveillance applications I was working on. I rejected Lucene at first only because of it being written in java and I personally felt something like C++ was a better fit for the high dimension job (no garbage collection to worry about). So I wrote a [high dimension indexer for MongoDB](https://jira.mongodb.org/browse/SERVER-3551) inspired by RTree (for the record it's implementation is based on XTree) and wrote it using C++ 14 preview features (lambda functions were the new hotness on the block and java didn't even have them yet). Even in C++ back then SIMD wasn't very well supported by the compiler natively so I had to add all sorts of compiler tricks to squeeze every ounce of vector parallelization to make it performant. C++ has gotten better since then but I think java still lags in this area? Even JEP 426 is a ways off (maybe because OpenJDK is holding these things hostage)? So maybe java is still not the right fit here? I wonder though, does that mean Lucene shouldn't provide dimensionality higher than arbitrary 1024? Maybe not. I agree dimensional reduction techniques like PCA should be considered to reduce the storage volume. The problem with that argument is that dimensionality reduction fails when features are weakly correlated. You can't capture the majority of the signal in the first  N components and therefore need higher dimensionality. But does that mean that 1024 is still too low to make Lucene a viable option? 

Aside from conjecture does anyone have empirical examples where 1024 is too low and what specific Lucene capabilities (e.g., scoring?) would make adding support for dimensions higher than 1024 really worth considering over using dimensionality reduction? If Lucene doesn't do this does it really risk the project becoming irrelevant? That sounds a bit like sensationalism. Even if higher dimensionality is added to the current vector implementation (I'd actually argue we should explore converting BKD to support higher dimensions instead) are we convinced it will reasonably perform without JEP 426 or better SIMD support that's only available in newer JDKs? Can anyone smart here post their benchmarks to substantiate their claims? I know Pinecone (and others) [have blogged about their love for RUST](https://www.pinecone.io/learn/rust-rewrite/) for these kinds of applications. Should Lucene just leave this to the job of alternative Search APIs? Maybe something like Tantivy or Rucene? Or is it time we explore a new optional Lucene Vector module that supports cutting edge JDK features through gradle tooling for optimizing the vector use case?

Interested what others think.

**Comment by mikemccand** [comment]
> Nice little rewrite ChatGPT did there.

❗ ❗ ❗

**Comment by dan-caprioara** [comment]
I also vote for the increase of dimensions. I use GPT Ada2 for embeddings and I have to recompile Lucene just to have the constant value increased. Having a system property for changing it would be nice.

**Comment by mauvo** [comment]
Just adding a perspective, we're evaluating options for indexing vectors and need to support [OpenAI's second-generation text embedding](https://platform.openai.com/docs/guides/embeddings/second-generation-models) that outputs 1536 dimensions. 

Raising Lucene's max to something like 2048 makes a lot of sense to me. Things are moving very quickly in the LLM space at the moment. I can see there are concerns about performance and scale, but today people don't choose Lucene to index vectors when they need to leverage GPU performance and have massive amounts of vector data. Anyway, just my two cents. :-)

**Comment by uschindler** [comment]
Here's an example why making the vector dimensions configurable is a bad idea: #12281
This issue shows that each added dimension makes the floating point errors larger and sometimes also returns NaN. Do we have tests when multiplying vectors causes NaN?

**Comment by alessandrobenedetti** [comment]
Copying and pasting here, just for visibility:
"> Here's an example why making the vector dimensions configurable is a bad idea: #12281 This issue shows that each added dimension makes the floating point errors larger and sometimes also returns NaN. Do we have tests when multiplying vectors causes NaN?

I may sound like somebody who contradicts another just for the sake of doing so, but I do genuinely believe these kind of discoveries support the fact that making it configurable is actually a good idea:
We are not changing a production system here but we are changing a library.
Enabling more users to experiment with higher dimensions increase the probability of finding (and then solving) this sort of issues.
I suspect we are not recommending anywhere here to go to prod with un-tested and un-benchmarked vector sizes anyway
"

**Comment by uschindler** [comment]
> Enabling more users to experiment with higher dimensions increase the probability of finding (and then solving) this sort of issues.

It also shows that this causes a long-tail of issues:
- If we would fix the mentioned issue in the way proposed there, the performance would be going down, as a widening conversion to double would break hotspot optimizations and would also not be possible with SIMD in a performant manner. So this is a reason to for now stay with the current limit.

**Comment by uschindler** [comment]
In addition if we raise the number of dimensions people will then start claiming for higher precision in calculations, completely forgetting that Lucene is a full text search engine to bring results in milliseconds not 2 hours. Score calculations introduce rounding anyways and making them exact is (a) not needed for lucene (we just sort on those values) and (b) would slow down the whole thing so much.

So keep with the current limit and NOT make it configurable. I agree to raise the maximum to 2048 (while recommending to people to use Java 20 for running Lucene and enable incubator vectors).

At same time close any issues about calculation precission and on the other hand get the JDK people support half float calculations.

**Comment by dsmiley** [comment]
I think a library should _empower_ a user to discover what works (and doesn't) for them, rather than playing big brother and insist it knows best that there's no way some high setting could ever work for any user.  Right?  By making it a system property that does not need to be configured for <= 1024, it should raise a red flag to users that they are venturing into unusual territory.  i.e. they've been warned.  They'd have to go looking for such a setting and see warnings; it's not something a user would do accidentally either.

> if we raise the number of dimensions people will then start claiming for higher precision in calculations, 

LOL People may ask for whatever they want :-) including using/abusing a system beyond its intended scope.  So what?  BTW I've thoroughly enjoyed seeing several use cases of my code in Lucene/Solr that I had never considered yet worked really well for a user :-D.  Pure joy.  Of course not every request makes sense to us.  I'd rather endure such than turn users away from Lucene that we can support trivially today.

**Comment by alessandrobenedetti** [comment]
> In addition if we raise the number of dimensions people will then start claiming for higher precision in calculations, completely forgetting that Lucene is a full text search engine to bring results in milliseconds not 2 hours. Score calculations introduce rounding anyways and making them exact is (a) not needed for lucene (we just sort on those values) and (b) would slow down the whole thing so much.
> 
> So keep with the current limit and NOT make it configurable. I agree to raise the maximum to 2048 (while recommending to people to use Java 20 for running Lucene and enable incubator vectors).
> 
> At same time close any issues about calculation precission and on the other hand get the JDK people support half float calculations.

@uschindler , I am not convinced but it's fine to have different opinions!
I do agree we should improve all the improvable and at the same time, in parallel, give users flexibility to experiment:
- break it
- make it super slow with enormous vectors
- make it super slow with enormous field content
- make it super slow with a great number of fields
...
I just think that the more you let your users do (with a reasonable effort), the more we'll gain users and consequentially improvements (it's a fact that more people you involve in a project as users the more you increase the probability of developing contributors and contributions).

We may have different opinions here and that's fine, but my intent as a committer is to build the best solution for the community rather than the best solution according to my ideas.

You know, if we wanted sub-ms responses all the time we could set a hard limit to 1024 chars per textual field and allow a very low number of fields, but then would Lucene attract any user at all?

**Comment by mayya-sharipova** [comment]
I would like to renew the issue in light of the recent [integration of incubating Panama Vector API](https://github.com/apache/lucene/pull/12311), as indexing of vectors with it much faster.

We run a benchmarking test, and indexing a dataset of vectors of 1536 dims was slightly faster than indexing of 1024 dims. This gives us enough confidence to extend max dims to 2048.

### Test environment 
- Dataset: 
  - [nq](https://huggingface.co/datasets/BeIR/nq) dataset with `text` field embedded with OpenAI `text-embedding-ada-002` model, 1536 dims

- [KnnGraphTester](https://github.com/apache/lucene/blob/branch_9_7/lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java)
-  maxConn: 16, beamWidthIndex: 100
- Apple M1 laptop

### Test1:
- Lucene 9.7 branch
- Panama Vector API not enabled
- vector dims=1024 (OpenAi vectors that were cut off to first 1024 dims)
- Results: Indexed 2680961 documents in **3287s**

<details>
 <summary>Details</summary>

```
 java -cp  "lib/*:classes" -Xmx16g -Xms16g org.apache.lucene.util.hnsw.KnnGraphTester -dim 1024 -ndoc 2680961 -reindex -docs vectors_dims1024.bin -maxConn 16 -beamWidthIndex 100
creating index in vectors_dims1024.bin-16-100.index
MS 0 [2023-06-26T11:10:24.765857Z; main]: initDynamicDefaults maxThreadCount=4 maxMergeCount=9
IFD 0 [2023-06-26T11:10:24.782017Z; main]: init: current segments file is "segments"; deletionPolicy=org.apache.lucene.index.KeepOnlyLastCommitDeletionPolicy@646d64ab
IFD 0 [2023-06-26T11:10:24.783554Z; main]: now delete 0 files: []
IFD 0 [2023-06-26T11:10:24.784291Z; main]: now checkpoint "" [0 segments ; isCommit = false]
IFD 0 [2023-06-26T11:10:24.784338Z; main]: now delete 0 files: []
IFD 0 [2023-06-26T11:10:24.785377Z; main]: 0 ms to checkpoint
IW 0 [2023-06-26T11:10:24.785523Z; main]: init: create=true reader=null
IW 0 [2023-06-26T11:10:24.790087Z; main]:
dir=MMapDirectory@/Users/mayya/Elastic/knn/open_ai_vectors/vectors_dims1024.bin-16-100.index lockFactory=org.apache.lucene.store.NativeFSLockFactory@2c039ac6
index=
version=9.7.0
analyzer=org.apache.lucene.analysis.standard.StandardAnalyzer
ramBufferSizeMB=1994.0
maxBufferedDocs=-1
mergedSegmentWarmer=null
delPolicy=org.apache.lucene.index.KeepOnlyLastCommitDeletionPolicy
commit=null
openMode=CREATE
similarity=org.apache.lucene.search.similarities.BM25Similarity
mergeScheduler=ConcurrentMergeScheduler: maxThreadCount=4, maxMergeCount=9, ioThrottle=true
codec=Lucene95
infoStream=org.apache.lucene.util.PrintStreamInfoStream
mergePolicy=[TieredMergePolicy: maxMergeAtOnce=10, maxMergedSegmentMB=5120.0, floorSegmentMB=2.0, forceMergeDeletesPctAllowed=10.0, segmentsPerTier=10.0, maxCFSSegmentSizeMB=8.796093022208E12, noCFSRatio=0.1, deletesPctAllowed=20.0
readerPooling=true
perThreadHardLimitMB=1945
useCompoundFile=false
commitOnClose=true
indexSort=null
checkPendingFlushOnUpdate=true
softDeletesField=null
maxFullFlushMergeWaitMillis=500
leafSorter=null
eventListener=org.apache.lucene.index.IndexWriterEventListener$1@2173f6d9
writer=org.apache.lucene.index.IndexWriter@307f6b8c

IW 0 [2023-06-26T11:10:24.790232Z; main]: MMapDirectory.UNMAP_SUPPORTED=true
DWPT 0 [2023-06-26T11:19:47.652040Z; main]: flush postings as segment _0 numDocs=460521
IW 0 [2023-06-26T11:19:47.653761Z; main]: 1 ms to write norms
IW 0 [2023-06-26T11:19:47.653954Z; main]: 0 ms to write docValues
IW 0 [2023-06-26T11:19:47.654032Z; main]: 0 ms to write points
IW 0 [2023-06-26T11:19:49.152263Z; main]: 1498 ms to write vectors
IW 0 [2023-06-26T11:19:49.166472Z; main]: 14 ms to finish stored fields
IW 0 [2023-06-26T11:19:49.166642Z; main]: 0 ms to write postings and finish vectors
IW 0 [2023-06-26T11:19:49.167167Z; main]: 0 ms to write fieldInfos
DWPT 0 [2023-06-26T11:19:49.167954Z; main]: new segment has 0 deleted docs
DWPT 0 [2023-06-26T11:19:49.168030Z; main]: new segment has 0 soft-deleted docs
DWPT 0 [2023-06-26T11:19:49.169572Z; main]: new segment has no vectors; no norms; no docValues; no prox; freqs
DWPT 0 [2023-06-26T11:19:49.169670Z; main]: flushedFiles=[_0_Lucene95HnswVectorsFormat_0.vem, _0.fdm, _0_Lucene95HnswVectorsFormat_0.vec, _0.fdx, _0_Lucene95HnswVectorsFormat_0.vex, _0.fdt, _0.fnm]
....
Indexed 2680961 documents in 3287s
```
</details>


### Test2
- Lucene 9.7 branch with FloatVectorValues.MAX_DIMENSIONS set to 2048
- Panama Vector API enabled
- dims=1536
- Results: Indexed 2680961 documents in **3141s** 

<details>
 <summary>Details</summary>

```
java --add-modules jdk.incubator.vector -cp  "lib/*:classes" -Xmx16g -Xms16g org.apache.lucene.util.hnsw.KnnGraphTester -dim 1536 -ndoc 2680961 -reindex -docs vectors.bin -maxConn 16 -beamWidthIndex 100

WARNING: Using incubator modules: jdk.incubator.vector
creating index in vectors.bin-16-100.index
Jun 26, 2023 10:34:29 A.M. org.apache.lucene.store.MemorySegmentIndexInputProvider <init>
INFO: Using MemorySegmentIndexInput with Java 20; to disable start with -Dorg.apache.lucene.store.MMapDirectory.enableMemorySegments=false
MS 0 [2023-06-26T14:34:29.271516Z; main]: initDynamicDefaults maxThreadCount=4 maxMergeCount=9
IFD 0 [2023-06-26T14:34:29.329779Z; main]: init: current segments file is "segments"; deletionPolicy=org.apache.lucene.index.KeepOnlyLastCommitDeletionPolicy@64f6106c
IFD 0 [2023-06-26T14:34:29.336415Z; main]: now delete 0 files: []
IFD 0 [2023-06-26T14:34:29.338546Z; main]: now checkpoint "" [0 segments ; isCommit = false]
IFD 0 [2023-06-26T14:34:29.338654Z; main]: now delete 0 files: []
IFD 0 [2023-06-26T14:34:29.347243Z; main]: 2 ms to checkpoint
IW 0 [2023-06-26T14:34:29.348255Z; main]: init: create=true reader=null
IW 0 [2023-06-26T14:34:29.368686Z; main]:
dir=MMapDirectory@/Users/mayya/Elastic/knn/open_ai_vectors/vectors.bin-16-100.index lockFactory=org.apache.lucene.store.NativeFSLockFactory@319b92f3
index=
version=9.7.0
analyzer=org.apache.lucene.analysis.standard.StandardAnalyzer
ramBufferSizeMB=1994.0
maxBufferedDocs=-1
mergedSegmentWarmer=null
delPolicy=org.apache.lucene.index.KeepOnlyLastCommitDeletionPolicy
commit=null
openMode=CREATE
similarity=org.apache.lucene.search.similarities.BM25Similarity
mergeScheduler=ConcurrentMergeScheduler: maxThreadCount=4, maxMergeCount=9, ioThrottle=true
codec=Lucene95
infoStream=org.apache.lucene.util.PrintStreamInfoStream
mergePolicy=[TieredMergePolicy: maxMergeAtOnce=10, maxMergedSegmentMB=5120.0, floorSegmentMB=2.0, forceMergeDeletesPctAllowed=10.0, segmentsPerTier=10.0, maxCFSSegmentSizeMB=8.796093022208E12, noCFSRatio=0.1, deletesPctAllowed=20.0
readerPooling=true
perThreadHardLimitMB=1945
useCompoundFile=false
commitOnClose=true
indexSort=null
checkPendingFlushOnUpdate=true
softDeletesField=null
maxFullFlushMergeWaitMillis=500
leafSorter=null
eventListener=org.apache.lucene.index.IndexWriterEventListener$1@10a035a0
writer=org.apache.lucene.index.IndexWriter@67b467e9

IW 0 [2023-06-26T14:34:29.369224Z; main]: MMapDirectory.UNMAP_SUPPORTED=true
Jun 26, 2023 10:34:29 A.M. org.apache.lucene.util.VectorUtilPanamaProvider <init>
INFO: Java vector incubator API enabled; uses preferredBitSize=128
DWPT 0 [2023-06-26T14:40:36.945965Z; main]: flush postings as segment _0 numDocs=314897
IW 0 [2023-06-26T14:40:36.949748Z; main]: 2 ms to write norms
IW 0 [2023-06-26T14:40:36.950336Z; main]: 0 ms to write docValues
IW 0 [2023-06-26T14:40:36.950452Z; main]: 0 ms to write points
IW 0 [2023-06-26T14:40:38.639069Z; main]: 1688 ms to write vectors
IW 0 [2023-06-26T14:40:38.669749Z; main]: 29 ms to finish stored fields
IW 0 [2023-06-26T14:40:38.670044Z; main]: 0 ms to write postings and finish vectors
IW 0 [2023-06-26T14:40:38.670847Z; main]: 0 ms to write fieldInfos
DWPT 0 [2023-06-26T14:40:38.672893Z; main]: new segment has 0 deleted docs
DWPT 0 [2023-06-26T14:40:38.673016Z; main]: new segment has 0 soft-deleted docs
DWPT 0 [2023-06-26T14:40:38.675915Z; main]: new segment has no vectors; no norms; no docValues; no prox; freqs
DWPT 0 [2023-06-26T14:40:38.676120Z; main]: flushedFiles=[_0_Lucene95HnswVectorsFormat_0.vem, _0.fdm, _0_Lucene95HnswVectorsFormat_0.vec, _0.fdx, _0_Lucene95HnswVectorsFormat_0.vex, _0.fdt, _0.fnm]
DWPT 0 [2023-06-26T14:40:38.676311Z; main]: flushed codec=Lucene95
DWPT 0 [2023-06-26T14:40:38.677609Z; main]: flushed: segment=_0 ramUsed=1,945.012 MB newFlushedSize=1,863.46 MB docs/MB=168.985
DWPT 0 [2023-06-26T14:40:38.680696Z; main]: flush time 1735.77025 ms
IW 0 [2023-06-26T14:40:38.682741Z; main]: publishFlushedSegment seg-private updates=null
IW 0 [2023-06-26T14:40:38.683738Z; main]: publishFlushedSegment _0(9.7.0):C314897:[diagnostics={source=flush, lucene.version=9.7.0, os.version=13.2.1, os.arch=x86_64, os=Mac OS X, java.vendor=Oracle Corporation, java.runtime.version=20.0.1+9-29, timestamp=1687790438678}]:[attributes={Lucene90StoredFieldsFormat.mode=BEST_SPEED}] :id=717x28qrd00q2ke3d17eerf4x
BD 0 [2023-06-26T14:40:38.687864Z; main]: finished packet delGen=1 now completedDelGen=1
IW 0 [2023-06-26T14:40:38.691420Z; main]: publish sets newSegment delGen=1 seg=_0(9.7.0):C314897:[diagnostics={source=flush, lucene.version=9.7.0, os.version=13.2.1, os.arch=x86_64, os=Mac OS X, java.vendor=Oracle Corporation, java.runtime.version=20.0.1+9-29, timestamp=1687790438678}]:[attributes={Lucene90StoredFieldsFormat.mode=BEST_SPEED}] :id=717x28qrd00q2ke3d17eerf4x
IFD 0 [2023-06-26T14:40:38.692639Z; main]: now checkpoint "_0(9.7.0):C314897:[diagnostics={source=flush, lucene.version=9.7.0, os.version=13.2.1, os.arch=x86_64, os=Mac OS X, java.vendor=Oracle Corporation, java.runtime.version=20.0.1+9-29, timestamp=1687790438678}]:[attributes={Lucene90StoredFieldsFormat.mode=BEST_SPEED}] :id=717x28qrd00q2ke3d17eerf4y" [1 segments ; isCommit = false]
IFD 0 [2023-06-26T14:40:38.693268Z; main]: now delete 0 files: []
IFD 0 [2023-06-26T14:40:38.693464Z; main]: 1 ms to checkpoint
MP 0 [2023-06-26T14:40:38.700301Z; main]:   seg=_0(9.7.0):C314897:[diagnostics={source=flush, lucene.version=9.7.0, os.version=13.2.1, os.arch=x86_64, os=Mac OS X, java.vendor=Oracle Corporation, java.runtime.version=20.0.1+9-29, timestamp=1687790438678}]:[attributes={Lucene90StoredFieldsFormat.mode=BEST_SPEED}] :id=717x28qrd00q2ke3d17eerf4y size=1863.460 MB
MP 0 [2023-06-26T14:40:38.701368Z; main]: findMerges: 1 segments
MP 0 [2023-06-26T14:40:38.701645Z; main]:   allowedSegmentCount=10 vs count=1 (eligible count=1)   
 ...

Indexed 2680961 documents in 3141s
```
</details>

**Comment by mikemccand** [comment]
> We run a benchmarking test, and indexing a dataset of vectors of 1536 dims was slightly faster than indexing of 1024 dims. This gives us enough confidence to extend max dims to 2048 (at least when vectorization is enabled).

I found this very strange at first :)

But then I read more closely, and I think what you meant is indexing 1024 dims without Panama (SIMD vector instructions) is slower than indexing 1536 dims with Panama enabled?  Which is really quite impressive.

Do we know what gains we see at search time going from 1024 -> 1536?

**Comment by mayya-sharipova** [comment]
@mikemccand Indeed, exactly as said, sorry for being unclear. We have not checked search, will work on that.

@uschindler Thanks, indeed, we need tests on other machines. +1 for raising dims to 2048 in HNSW codec.

**Comment by rmuir** [comment]
Can we run this test with lucene's defaults (e.g. not a 2GB rambuffer)? 
We are still talking about an hour to index < 3M docs, so I think the performance is not good.
As i've said before, i never thought 1024 was a good situation either. 768 is also excruciating. 
Purpose of the vectorization is just to alleviate some of the pain. It is like giving the patient an aspirin, it doesn't really fix the problem.

**Comment by alessandrobenedetti** [comment]
I am extremely curious, what should we consider a good performance to index <3M docs?
I mean, I agree we should always try to improve things and aim for the stars, but as maintainers of a library who are we to decide what's acceptable and what's not for the users?
Is it because of a comparison with other libraries or solutions?
They may have many reasons for being faster (and definitely we should take inspiration)
If we look to : https://home.apache.org/~mikemccand/lucenebench/indexing.html , we clearly improved the indexing throughput substantially over the years, does this mean that Lucene back in 2011 should have not committed additional features/improvements because for some people (people from the future) "it was slow"?

**Comment by dweiss** [comment]
Leaving a higher number of segments dodges the merge costs, I think.

**Comment by jpountz** [comment]
This benchmark really only measures the flushing cost, as `ConcurrentMergeScheduler` is used, so merges run in background threads. So the improvement makes sense to me as the cost of adding vectors into a HNSW graph increases as the size of the HNSW graph increases. If we want to get a sense of the number of docs per second per core that we support with a 2GB RAM buffer vs. the 16MB default, using `SerialMergeScheduler` would be a better choice.

**Comment by uschindler** [comment]
Hi,
actually this issue is already resolved, although the DEFAULT did not change (and won't change due to performance risks), see here: https://github.com/apache/lucene/pull/12436 - this PR allows users of Lucene to raise the limit (at least for HNSW codec) on codec level.

To implement (on your own risk), create your own ` KnnVectorsFormat` and let it return a different number fog `getMaxDimensions()`. Then construct your own codec from it and index your data.

**Comment by uschindler** [comment]
@mayya-sharipova: Should we close this issue or are there any plans to also change the default maximum? I don't think so.

**Comment by mayya-sharipova** [comment]
Yes, thanks for the reminder. Now Codec is responsible for managing dims, we can close it.

**Comment by jzwolak** [comment]
@asfimport

I suggest making it easier to change the limit. I appreciate the need to have a limit for optimization and performance. The models that have more than 1024 dimensions in their embedding vectors are not bad models or poor choices. In fact, there are models that have millions of dimensions. And there's real value there. Though I will add that the model I heard about at a conference talk with millions of dimensions did simplify the vector to have base-2 integers instead of floating point numbers. That isn't compatible with what Lucene is doing here, but my point remains that there are larger models and valid reasoning for such models.

I'm working with a model with 3072 dimensions and getting _great_ results. Reducing the dimensions will lose information.

Last, other vector search indexers/databases support very quick search for 3072 dimensions. Like Qdrant.

**Comment by mkhludnev** [comment]
fwiw 

https://openai.com/index/new-embedding-models-and-api-updates/

> text-embedding-3-large embedding can be shortened to a size of 256 while still outperforming an unshortened text-embedding-ada-002 embedding with a size of 1536.

It's about Matryoshka Representation Learning (MRL)
